modules/sfm/CMakeLists.txt

```diff
@@ -3,23 +3,56 @@ set(the_description "SFM algorithms")
 
 
 ### LIBMV LIGHT EXTERNAL DEPENDENCIES ###
-list(APPEND CMAKE_MODULE_PATH "${CMAKE_CURRENT_LIST_DIR}/cmake")
-find_package(Gflags QUIET)
+
 find_package(Ceres QUIET)
-if(NOT Ceres_FOUND)  # Looks like Ceres find glog on the own, so separate search isn't necessary
+
+if(NOT Gflags_FOUND)  # Ceres find gflags on the own, so separate search isn't necessary
+  find_package(Gflags QUIET)
+endif()
+if(NOT Glog_FOUND)  # Ceres find glog on the own, so separate search isn't necessary
   find_package(Glog QUIET)
 endif()
 
-if((gflags_FOUND OR GFLAGS_FOUND OR GFLAGS_INCLUDE_DIRS) AND (glog_FOUND OR GLOG_FOUND OR GLOG_INCLUDE_DIRS))
-  set(_fname "${CMAKE_CURRENT_BINARY_DIR}/test_sfm_deps.cpp")
-  file(WRITE "${_fname}" "#include <glog/logging.h>\n#include <gflags/gflags.h>\nint main() { (void)(0); return 0; }\n")
-  try_compile(SFM_DEPS_OK "${CMAKE_BINARY_DIR}" "${_fname}"
-      CMAKE_FLAGS "-DINCLUDE_DIRECTORIES:STRING=${GLOG_INCLUDE_DIRS};${GFLAGS_INCLUDE_DIRS}"
-      LINK_LIBRARIES ${GLOG_LIBRARIES} ${GFLAGS_LIBRARIES}
-      OUTPUT_VARIABLE OUTPUT
-  )
-  file(REMOVE "${_fname}")
-  message(STATUS "Checking SFM deps... ${SFM_DEPS_OK}")
+if(NOT Gflags_FOUND OR NOT Glog_FOUND)
+  # try local search scripts
+  list(APPEND CMAKE_MODULE_PATH "${CMAKE_CURRENT_LIST_DIR}/cmake")
+  if(NOT Gflags_FOUND)
+    find_package(Gflags QUIET)
+  endif()
+  if(NOT Glog_FOUND)
+    find_package(Glog QUIET)
+  endif()
+endif()
+
+if(NOT DEFINED GFLAGS_INCLUDE_DIRS AND DEFINED GFLAGS_INCLUDE_DIR)
+  set(GFLAGS_INCLUDE_DIRS "${GFLAGS_INCLUDE_DIR}")
+endif()
+if(NOT DEFINED GLOG_INCLUDE_DIRS AND DEFINED GLOG_INCLUDE_DIR)
+  set(GLOG_INCLUDE_DIRS "${GLOG_INCLUDE_DIR}")
+endif()
+
+if((gflags_FOUND OR Gflags_FOUND OR GFLAGS_FOUND OR GFLAGS_INCLUDE_DIRS) AND (glog_FOUND OR Glog_FOUND OR GLOG_FOUND OR GLOG_INCLUDE_DIRS))
+  set(__cache_key "${GLOG_INCLUDE_DIRS} ~ ${GFLAGS_INCLUDE_DIRS} ~ ${GLOG_LIBRARIES} ~ ${GFLAGS_LIBRARIES}")
+  if(NOT DEFINED SFM_GLOG_GFLAGS_TEST_CACHE_KEY OR NOT (SFM_GLOG_GFLAGS_TEST_CACHE_KEY STREQUAL __cache_key))
+    set(__fname "${CMAKE_CURRENT_LIST_DIR}/cmake/checks/check_glog_gflags.cpp")
+    try_compile(
+        SFM_GLOG_GFLAGS_TEST "${CMAKE_BINARY_DIR}" "${__fname}"
+        CMAKE_FLAGS "-DINCLUDE_DIRECTORIES:STRING=${GLOG_INCLUDE_DIRS};${GFLAGS_INCLUDE_DIRS}"
+        LINK_LIBRARIES ${GLOG_LIBRARIES} ${GFLAGS_LIBRARIES}
+        OUTPUT_VARIABLE __output
+    )
+    if(NOT SFM_GLOG_GFLAGS_TEST)
+      file(APPEND ${CMAKE_BINARY_DIR}${CMAKE_FILES_DIRECTORY}/CMakeError.log
+          "Failed compilation check: ${__fname}\n"
+          "${__output}\n\n"
+      )
+    endif()
+    set(SFM_GLOG_GFLAGS_TEST "${SFM_GLOG_GFLAGS_TEST}" CACHE INTERNAL "")
+    set(SFM_GLOG_GFLAGS_TEST_CACHE_KEY "${__cache_key}" CACHE INTERNAL "")
+    message(STATUS "Checking SFM glog/gflags deps... ${SFM_GLOG_GFLAGS_TEST}")
```

Comment:
@LaurentBerger 

> new.txt
> -- Checking SFM deps... FALSE

Logs don't correspond to the content of this patch.
Please re-check that patch is applied properly.

Reply:
`
Failed compilation check: G:/Lib/opencv_contrib/modules/sfm/cmake/checks/check_glog_gflags.cpp
Change Dir: G:/Lib/build/opencv/CMakeFiles/CMakeTmp

Run Build Command(s):C:/Program Files (x86)/Microsoft Visual Studio/2017/Enterprise/MSBuild/15.0/Bin/MSBuild.exe cmTC_32546.vcxproj /p:Configuration=Debug /p:Platform=x64 /p:VisualStudioVersion=15.0 /v:m && Microsoft (R) Build Engine version 15.9.21+g9802d43bc3 pour .NET Framework

Copyright (C) Microsoft Corporation. Tous droits réservés.



  Compilateur d'optimisation Microsoft (R) C/C++ version 19.16.27041 pour x64

  Copyright (C) Microsoft Corporation. Tous droits réservés.

  

  cl /c /IG:\Lib\install\gflags\include /Zi /W4 /WX- /diagnostics:classic /MP /Od /Ob0 /Oi /D WIN32 /D _WINDOWS /D _CRT_SECURE_NO_DEPRECATE /D _CRT_NONSTDC_NO_DEPRECATE /D _SCL_SECURE_NO_WARNINGS /D GFLAGS_IS_A_DLL=1 /D "CMAKE_INTDIR=\"Debug\"" /D _MBCS /Gm- /EHa /RTC1 /MDd /GS /Gy /fp:precise /Zc:wchar_t /Zc:forScope /Zc:inline /GR /Fo"cmTC_32546.dir\Debug\\" /Fd"cmTC_32546.dir\Debug\vc141.pdb" /Gd /TP /wd4127 /wd4251 /wd4324 /wd4275 /wd4512 /wd4589 /errorReport:queue  /bigobj G:\Lib\opencv_contrib\modules\sfm\cmake\checks\check_glog_gflags.cpp

  check_glog_gflags.cpp

  

G:\Lib\opencv_contrib\modules\sfm\cmake\checks\check_glog_gflags.cpp(1): fatal error C1083: Impossible d'ouvrir le fichier include : 'glog/logging.h' : No such file or directory [G:\Lib\build\opencv\CMakeFiles\CMakeTmp\cmTC_32546.vcxproj]
`
Is it good message ?

Reply:
I see `/IG:\Lib\install\gflags\include` but don't see include directory for glog from this log.

Log above requests to specify `GLOG_INCLUDE_DIR`:

```
-- Failed to find glog - Could not find glog include directory, set GLOG_INCLUDE_DIR to directory containing glog/logging.h
```

Could you dump used CMake options?

Reply:
OK I added : 
```
-DEigen3_DIR:PATH=${installRepo}/Eigen3/share/eigen3/cmake \
-Dglog_DIR:PATH=${installRepo}/glog/lib/cmake/glog -Dgflags_DIR:PATH=${installRepo}/gflags/lib/cmake/gflags \
-DGLOG_INCLUDE_DIR=${installRepo}/glog/include \

```
and sfm is true : 
```

-- Julia not found. Not compiling Julia Bindings.
-- Module opencv_ovis disabled because OGRE3D was not found
-- No preference for use of exported gflags CMake configuration set, and no hints for include/library directories provided. Defaulting to preferring an installed/exported gflags CMake configuration if available.
-- Found installed version of gflags: G:/lib/install/gflags/lib/cmake/gflags
-- Detected gflags version: 2.2.1
CMake Warning (dev) at C:/Program Files/CMake/share/cmake-3.18/Modules/FindPackageHandleStandardArgs.cmake:273 (message):
  The package name passed to `find_package_handle_standard_args` (Gflags)
  does not match the name of the calling package (gflags).  This can lead to
  problems in calling code that expects `find_package` result variables
  (e.g., `_FOUND`) to follow a certain pattern.
Call Stack (most recent call first):
  G:/Lib/install/ceres-solver/CMake/FindGflags.cmake:581 (find_package_handle_standard_args)
  G:/Lib/install/ceres-solver/CMake/CeresConfig.cmake:256 (find_package)
  G:/Lib/opencv_contrib/modules/rgbd/CMakeLists.txt:3 (find_package)
This warning is for project developers.  Use -Wno-dev to suppress it.

-- No preference for use of exported gflags CMake configuration set, and no hints for include/library directories provided. Defaulting to preferring an installed/exported gflags CMake configuration if available.
-- Found installed version of gflags: G:/lib/install/gflags/lib/cmake/gflags
-- Detected gflags version: 2.2.1
CMake Warning (dev) at C:/Program Files/CMake/share/cmake-3.18/Modules/FindPackageHandleStandardArgs.cmake:273 (message):
  The package name passed to `find_package_handle_standard_args` (Gflags)
  does not match the name of the calling package (gflags).  This can lead to
  problems in calling code that expects `find_package` result variables
  (e.g., `_FOUND`) to follow a certain pattern.
Call Stack (most recent call first):
  G:/Lib/install/ceres-solver/CMake/FindGflags.cmake:581 (find_package_handle_standard_args)
  G:/Lib/install/ceres-solver/CMake/CeresConfig.cmake:256 (find_package)
  G:/Lib/opencv_contrib/modules/sfm/CMakeLists.txt:7 (find_package)
This warning is for project developers.  Use -Wno-dev to suppress it.

-- Tesseract:   YES (ver unknown)
-- Allocator metrics storage type: 'long long'
-- Registering hook 'INIT_MODULE_SOURCES_opencv_dnn': G:/Lib/opencv/modules/dnn/cmake/hooks/INIT_MODULE_SOURCES_opencv_dnn.cmake
-- opencv_dnn: filter out cuda4dnn source code
-- No preference for use of exported gflags CMake configuration set, and no hints for include/library directories provided. Defaulting to preferring an installed/exported gflags CMake configuration if available.
-- Found installed version of gflags: G:/lib/install/gflags/lib/cmake/gflags
-- Detected gflags version: 2.2.1
CMake Warning (dev) at C:/Program Files/CMake/share/cmake-3.18/Modules/FindPackageHandleStandardArgs.cmake:273 (message):
  The package name passed to `find_package_handle_standard_args` (Gflags)
  does not match the name of the calling package (gflags).  This can lead to
  problems in calling code that expects `find_package` result variables
  (e.g., `_FOUND`) to follow a certain pattern.
Call Stack (most recent call first):
  G:/Lib/install/ceres-solver/CMake/FindGflags.cmake:581 (find_package_handle_standard_args)
  G:/Lib/install/ceres-solver/CMake/CeresConfig.cmake:256 (find_package)
  G:/Lib/opencv_contrib/modules/sfm/CMakeLists.txt:7 (find_package)
This warning is for project developers.  Use -Wno-dev to suppress it.

-- SYCL/OpenCL samples are skipped: SYCL SDK is required
--    - check configuration of SYCL_DIR/SYCL_ROOT/CMAKE_MODULE_PATH
--    - ensure that right compiler is selected from SYCL SDK (e.g, clang++): CMAKE_CXX_COMPILER=C:/Program Files (x86)/Microsoft Visual Studio/2017/Enterprise/VC/Tools/MSVC/14.16.27023/bin/Hostx64/x64/cl.exe
-- Pylint: registered 181 targets. Build 'check_pylint' target to run checks ("cmake --build . --target check_pylint" or "make check_pylint")
CMake Warning at cmake/OpenCVGenSetupVars.cmake:54 (message):
  CONFIGURATION IS NOT SUPPORTED: validate setupvars script in install
  directory
Call Stack (most recent call first):
  CMakeLists.txt:976 (include)


--
-- General configuration for OpenCV 4.5.0-dev =====================================
--   Version control:               4.4.0-473-g691c3d1e3c
--
--   Extra modules:
--     Location (extra):            G:/Lib/opencv_contrib/modules
--     Version control (extra):     4.1.1-410-gb65b3cc37-dirty
--
--   Platform:
--     Timestamp:                   2020-11-03T20:39:19Z
--     Host:                        Windows 10.0.19041 AMD64
--     CMake:                       3.18.4
--     CMake generator:             Visual Studio 15 2017 Win64
--     CMake build tool:            C:/Program Files (x86)/Microsoft Visual Studio/2017/Enterprise/MSBuild/15.0/Bin/MSBuild.exe
--     MSVC:                        1916
--
--   CPU/HW features:
--     Baseline:                    SSE SSE2 SSE3
--       requested:                 SSE3
--     Dispatched code generation:  SSE4_1 SSE4_2 FP16 AVX AVX2 AVX512_SKX
--       requested:                 SSE4_1 SSE4_2 AVX FP16 AVX2 AVX512_SKX
--       SSE4_1 (16 files):         + SSSE3 SSE4_1
--       SSE4_2 (2 files):          + SSSE3 SSE4_1 POPCNT SSE4_2
--       FP16 (1 files):            + SSSE3 SSE4_1 POPCNT SSE4_2 FP16 AVX
--       AVX (5 files):             + SSSE3 SSE4_1 POPCNT SSE4_2 AVX
--       AVX2 (30 files):           + SSSE3 SSE4_1 POPCNT SSE4_2 FP16 FMA3 AVX AVX2
--       AVX512_SKX (7 files):      + SSSE3 SSE4_1 POPCNT SSE4_2 FP16 FMA3 AVX AVX2 AVX_512F AVX512_COMMON AVX512_SKX
--
--   C/C++:
--     Built as dynamic libs?:      YES
--     C++ standard:                11
--     C++ Compiler:                C:/Program Files (x86)/Microsoft Visual Studio/2017/Enterprise/VC/Tools/MSVC/14.16.27023/bin/Hostx64/x64/cl.exe  (ver 19.16.27041.0)
--     C++ flags (Release):         /DWIN32 /D_WINDOWS /W4 /GR  /D _CRT_SECURE_NO_DEPRECATE /D _CRT_NONSTDC_NO_DEPRECATE /D _SCL_SECURE_NO_WARNINGS /Gy /bigobj /Oi  /fp:precise     /EHa /wd4127 /wd4251 /wd4324 /wd427 5 /wd4512 /wd4589 /MP  /MD /O2 /Ob2 /DNDEBUG
--     C++ flags (Debug):           /DWIN32 /D_WINDOWS /W4 /GR  /D _CRT_SECURE_NO_DEPRECATE /D _CRT_NONSTDC_NO_DEPRECATE /D _SCL_SECURE_NO_WARNINGS /Gy /bigobj /Oi  /fp:precise     /EHa /wd4127 /wd4251 /wd4324 /wd427 5 /wd4512 /wd4589 /MP  /MDd /Zi /Ob0 /Od /RTC1
--     C Compiler:                  C:/Program Files (x86)/Microsoft Visual Studio/2017/Enterprise/VC/Tools/MSVC/14.16.27023/bin/Hostx64/x64/cl.exe
--     C flags (Release):           /DWIN32 /D_WINDOWS /W3  /D _CRT_SECURE_NO_DEPRECATE /D _CRT_NONSTDC_NO_DEPRECATE /D _SCL_SECURE_NO_WARNINGS /Gy /bigobj /Oi  /fp:precise     /MP   /MD /O2 /Ob2 /DNDEBUG
--     C flags (Debug):             /DWIN32 /D_WINDOWS /W3  /D _CRT_SECURE_NO_DEPRECATE /D _CRT_NONSTDC_NO_DEPRECATE /D _SCL_SECURE_NO_WARNINGS /Gy /bigobj /Oi  /fp:precise     /MP /MDd /Zi /Ob0 /Od /RTC1
--     Linker flags (Release):      /machine:x64  /INCREMENTAL:NO
--     Linker flags (Debug):        /machine:x64  /debug /INCREMENTAL
--     ccache:                      NO
--     Precompiled headers:         YES
--     Extra dependencies:          opengl32 glu32
--     3rdparty dependencies:
--
--   OpenCV modules:
--     To be built:                 alphamat aruco bgsegm bioinspired calib3d ccalib core datasets dnn dnn_objdetect dnn_superres dpm face features2d flann fuzzy hfs highgui img_hash imgcodecs imgproc intensity_trans form line_descriptor mcc ml objdetect optflow phase_unwrapping photo plot python3 quality rapid reg saliency sfm shape stereo stitching structured_light superres surface_matching text tracking ts video videoio videos tab viz xfeatures2d ximgproc xobjdetect xphoto
--     Disabled:                    gapi rgbd world
--     Disabled by dependency:      -
--     Unavailable:                 cnn_3dobj cudaarithm cudabgsegm cudacodec cudafeatures2d cudafilters cudaimgproc cudalegacy cudaobjdetect cudaoptflow cudastereo cudawarping cudev cvv freetype hdf java js julia ma tlab ovis python2
--     Applications:                tests perf_tests examples apps
--     Documentation:               doxygen python
--     Non-free algorithms:         YES
--
--   Windows RT support:            NO
--
--   GUI:
--     Win32 UI:                    YES
--     OpenGL support:              YES (opengl32 glu32)
--     VTK support:                 YES (ver 9.0.20200702)
--
--   Media I/O:
--     ZLib:                        optimized G:/lib/install/zlib/lib/zlibstatic.lib debug G:/lib/install/zlib/lib/zlibstaticd.lib (ver 1.2.11)
--     JPEG:                        build-libjpeg-turbo (ver 2.0.5-62)
--     WEBP:                        build (ver encoder: 0x020f)
--     PNG:                         optimized G:/lib/install/libpng/lib/libpng16_static.lib debug G:/lib/install/libpng/lib/libpng16_staticd.lib (ver 1.6.36)
--     TIFF:                        build (ver 42 - 4.0.10)
--     JPEG 2000:                   build (ver 2.3.1)
--     OpenEXR:                     build (ver 2.3.0)
--     HDR:                         YES
--     SUNRASTER:                   YES
--     PXM:                         YES
--     PFM:                         YES
--
--   Video I/O:
--     DC1394:                      NO
--     FFMPEG:                      YES (prebuilt binaries)
--       avcodec:                   YES (58.91.100)
--       avformat:                  YES (58.45.100)
--       avutil:                    YES (56.51.100)
--       swscale:                   YES (5.7.100)
--       avresample:                YES (4.0.0)
--     GStreamer:                   NO
--     DirectShow:                  YES
--     Media Foundation:            YES
--       DXVA:                      YES
--
--   Parallel framework:            Concurrency
--
--   Other third-party libraries:
--     Intel IPP:                   2020.0.0 Gold [2020.0.0]
--            at:                   G:/Lib/build/opencv/3rdparty/ippicv/ippicv_win/icv
--     Intel IPP IW:                sources (2020.0.0)
--               at:                G:/Lib/build/opencv/3rdparty/ippicv/ippicv_win/iw
--     Lapack:                      NO
--     Eigen:                       YES (ver 3.3.90)
--     Custom HAL:                  NO
--     Protobuf:                    build (3.5.1)
--
--   OpenCL:                        YES (NVD3D11)
--     Include path:                G:/Lib/opencv/3rdparty/include/opencl/1.2
--     Link libraries:              Dynamic load
--
--   Python 3:
--     Interpreter:                 f:/Program Files/Python/python.exe (ver 3.8.5)
--     Libraries:                   optimized F:/Program Files/Python/libs/python38.lib debug F:/Program Files/Python/libs/python38_d.lib (ver 3.8.5)
--     numpy:                       f:/Program Files/Python/lib/site-packages/numpy/core/include (ver 1.18.5)
--     install path:                f:/Program Files/Python/lib/site-packages/cv2/python-3.8
--
--   Python (for build):            f:/Program Files/Python/python.exe
--     Pylint:                      F:/Program Files/Python/Scripts/pylint.exe (ver: 3.8.5, checks: 181)
--     Flake8:                      F:/Program Files/Python/Scripts/flake8.exe (ver: 3.8.3)
--
--   Java:
--     ant:                         NO
--     JNI:                         C:/Program Files/Java/jdk1.8.0_152/include C:/Program Files/Java/jdk1.8.0_152/include/win32 C:/Program Files/Java/jdk1.8.0_152/include
--     Java wrappers:               NO
--     Java tests:                  NO
--
--   Install to:                    G:/Lib/install/opencv
-- -----------------------------------------------------------------
--
-- Configuring done
-- Generating done
-- Build files have been written to: G:/Lib/build/opencv
*************************  -->devenv debug
Microsoft (R) Build Engine version 15.9.21+g9802d43bc3 pour .NET Framework
Copyright (C) Microsoft Corporation. Tous droits r▒serv▒s.
```

But now I have got an error building sfm : 

```
1>------ Build started: Project: opencv_sfm, Configuration: Release x64 ------
1>simple_pipeline.cpp
1>G:\lib\install\glog\include\glog/log_severity.h(51): fatal error C1189: #error:  ERROR macro is defined. Define GLOG_NO_ABBREVIATED_SEVERITIES before including logging.h. See the document for detail.
1>Done building project "opencv_sfm.vcxproj" -- FAILED.

```

Reply:
> fatal error C1189: #error:  ERROR macro is defined. Define GLOG_NO_ABBREVIATED_SEVERITIES before including logging.h. See the document for detail.

Added to handle that (not really related to Ceres 2.0.0):
```
add_definitions(/DGLOG_NO_ABBREVIATED_SEVERITIES)
```